agent/consul/state/stream_publisher.go

```diff
@@ -0,0 +1,439 @@
+		snapCache:       make(map[agentpb.Topic]map[string]*stream.EventSnapshot),
+		snapFns:         make(map[agentpb.Topic]stream.SnapFn),
+		subsByToken:     make(map[string]map[*agentpb.SubscribeRequest]*stream.Subscription),
+		commitCh:        make(chan commitUpdate, 64),
+	}
+
+	// create a local handler table
+	for topic, handlers := range topicRegistry {
+		fnCopy := handlers.Snapshot
+		e.snapFns[topic] = func(req *agentpb.SubscribeRequest, buf *stream.EventBuffer) (uint64, error) {
+			return fnCopy(e.store, req, buf)
+		}
+	}
+
+	go e.handleUpdates()
+
+	return e
+}
+
+// // PreparePublish accepts a set of events to be published. All events must have
+// // the same index which must be higher than any previously Committed events.
+// // Events will not be delivered to subscribers until Commit is called. If
+// // PreparePublish is called a second time before Commit, and the events have a
+// // higher index, it is assumed the previous set of events were aborted and they
+// // will be dropped. If the events in a subsequent call have the same index, they
+// // will be staged in addition to those already prepared and all will be
+// // committed on next Commit.
+// func (e *EventPublisher) PreparePublish(events []agentpb.Event) error {
+// 	if len(events) == 0 {
+// 		// no-op
+// 		return nil
+// 	}
+
+// 	e.stagedLock.Lock()
+// 	defer e.stagedLock.Unlock()
+
+// 	if len(e.staged) > 0 {
+// 		if e.staged[0].Index != events[0].Index {
+// 			// implicit abort - these events are from a different index, replace
+// 			// current staged events.
+// 			e.staged = events
+// 		} else {
+// 			// Merge events at same index
+// 			e.staged = append(e.staged, events...)
+// 		}
+// 	} else {
+// 		e.staged = events
+// 	}
+// 	return nil
+// }
+
+// // Commit triggers any staged events to be sent to the relevant listeners. This
+// // is called via txn.Defer to delay it from running until the transaction has
+// // been finalized.
+// func (e *EventPublisher) Commit() {
+// 	e.stagedLock.Lock()
+// 	defer e.stagedLock.Unlock()
+
+// 	if len(e.staged) == 0 {
+// 		return
+// 	}
+
+// 	e.commitCh <- commitUpdate{
+// 		tx:     e.store.db.Txn(false),
+// 		events: e.staged,
+// 	}
+// 	e.staged = nil
+// }
+
+func (e *EventPublisher) publishChanges(tx *txnWrapper, changes memdb.Changes) error {
+	var events []agentpb.Event
+	for topic, th := range topicRegistry {
+		if th.ProcessChanges != nil {
+			es, err := th.ProcessChanges(e.store, tx, changes)
+			if err != nil {
+				return fmt.Errorf("failed generating events for topic %q: %s", topic, err)
+			}
+			events = append(events, es...)
+		}
+	}
+	e.commitCh <- commitUpdate{
+		// Create a new transaction since it's going to be used from a different
+		// thread. Transactions aren't thread safe but it's OK to create it here
+		// since we won't try to use it in this thread and pass it straight to the
+		// handler which will own it exclusively.
+		tx:     e.store.db.Txn(false),
```

Comment:
I don't understand the "since it's going to be used from a different thread" part. Why is it necessary to create this transaction here instead of from the goroutine? I suspect if I knew more about memdb it might make more sense to me.

Does this transaction need to be opened before the "active transaction" (the one that produced the changeset) is committed?

Reply:
> Does this transaction need to be opened before the "active transaction" (the one that produced the changeset) is committed?

Yes it's necessary so that the event generating code is working with the exact same state of the store if it has to do any lookups that was true at the time the event was emitted. 

> Why is it necessary to create this transaction here instead of from the goroutine?
 
You basically figured it out with the question above - If we just passed execution to another goroutine here, returned from the FSM Apply and then let the publishing goroutine open a new transaction some time later, by the time it ran the state may have subsequently been changed and so the data it sees and publishes for an event would not be consistent with the state of the store at the time the event was actually emitted.

Originally, we forgot about thread safety entirely and passed the actual _same_ transaction across threads which caused lots of issues.

Another way to think about it is creating a read transaction like this is effectively taking a snapshot of the state store as it exists right now at the moment the event is published and passing it to be used by the publishing goroutine should it need to look up anything.

You picked up on the implicit part that should probably be added to the comment though. Something like:

```
// Create a new read-only transaction which becomes a snapshot of the state 
// at the time the Events were emitted that can be passed to the published in a 
// different goroutine. We can't pass the current active transaction as transaction 
// objects are not thread safe but we must open a new one here so that it is consistent 
// with the state at the point the events were emitted.
```

Reply:
Thanks! Sounds good